[SDL-0296] Feature/ possibility to update video streaming capabilities

yoooriii opened the pull request:
Fixes #1730

This PR is **ready** for review.

### Risk
This PR makes **minor** API changes.
This PR might affect other PRs in review: #1652 and #1613 

### Testing Plan
- [x] I have verified that I have not introduced new warnings in this PR (or explain why below)
- [x] I have run the unit tests with this PR
- [x] I have tested this PR against Core and verified behavior (if applicable, if not applicable, explain why below).

#### Unit Tests
Test cases in SDLStreamingVideoLifecycleManagerSpec reworked considerably.
Implemented new SDLCarWindowSpec test.

#### Core Tests

List of tests performed against Core and behaviours verified
Core version / branch / commit hash / module tested against: [Core: 0296_possibility_to_update_video_streaming_capabilities](https://github.com/LuxoftSDL/sdl_core/tree/feature/0296_possibility_to_update_video_streaming_capabilities)
HMI name / version / branch / commit hash / module tested against: [HMI: 0296_possibility_to_update_video_streaming_capabilities](https://github.com/LuxoftSDL/sdl_hmi/tree/feature/0296_possibility_to_update_video_streaming_capabilities)


### Summary

- Add restart of video service on VIDEO_STREAMING notification
- Implement view resize for streaming on new params
- Implement additional state for StreamingStateMachine
- Add callback for developers to provide streaming constraints


### CLA
- [x] I have signed [the CLA](https://docs.google.com/forms/d/e/1FAIpQLSdsgJY33VByaX482zHzi-xUm49JNnmuJOyAM6uegPQ2LXYVfA/viewform)

yoooriii commented on SmartDeviceLinkTests/ProxySpecs/SDLHapticManagerSpec.m:
```diff
@@ -504,6 +520,8 @@ @interface SDLFocusableItemLocator ()
                  hapticManager.videoScaleManager = [[SDLStreamingVideoScaleManager alloc] initWithScale:testUpdatedScale displayViewportResolution:testScreenSize];
                  [hapticManager updateInterfaceLayout];
+
+                 NSLog(@"%@", uiViewController);
```
TODO: remove NSLog

yoooriii commented:
The 2 tests SDLStreamingVideoLifecycleManagerSpec and SDLStreamingMediaManagerSpec failed. I am working on tests and the classes they test.

yoooriii commented:
Updated appCapability, lib --> head unit notification (was wrong json), updated video stream restart logic

SatbirTanda commented on Example Apps/Example Video ObjC/ProxyManager.h:
```diff
@@ -0,0 +1,41 @@
+
++ (instancetype)sharedManager;
+- (void)startWithProxyTransportType:(ProxyTransportType)proxyTransportType;
+- (void)startProxyTCP:(SDLTCPConfig*)tcpConfig;
```
`SDLTCPConfig *`

SatbirTanda commented on Example Apps/Example Video ObjC/SDLTCPConfig.m:
```diff
@@ -0,0 +1,24 @@
+
+@implementation SDLTCPConfig
+
++ (instancetype)configWithHost:(NSString*)host port:(UInt16)port {
```
`NSString *`

SatbirTanda commented on Example Apps/Example Video ObjC/SDLTCPConfig.m:
```diff
@@ -0,0 +1,24 @@
+    return [[self alloc] initWithHost:host port:port];
+}
+
+- (instancetype)initWithHost:(NSString*)host port:(UInt16)port {
```
`NSString *`

SatbirTanda commented on SmartDeviceLink-iOS.xcodeproj/project.pbxproj:
```diff
@@ -8929,13 +9227,16 @@
 				LD_RUNPATH_SEARCH_PATHS = "$(inherited) @executable_path/Frameworks";
 				MARKETING_VERSION = 6.6.0;
-				PRODUCT_BUNDLE_IDENTIFIER = com.smartdevicelink.SDLTestApp;
+				PRODUCT_BUNDLE_IDENTIFIER = pe.lo.smartdevicelink;
```
curious, why is this needed?

SatbirTanda commented on SmartDeviceLink-iOS.xcodeproj/project.pbxproj:
```diff
@@ -8913,13 +9208,16 @@
-				DEVELOPMENT_TEAM = NCVC2MHU7M;
+				CODE_SIGN_STYLE = Manual;
+				CURRENT_PROJECT_VERSION = 4;
+				DEVELOPMENT_TEAM = G99VP5HXV7;
```
You do not need to commit changes to this file; these changes seem specific for your environment

SatbirTanda commented on SmartDeviceLink/NSMapTable+Subscripting.m:
```diff
@@ -20,6 +20,10 @@ - (nullable id)objectForKeyedSubscript:(id<NSCopying>)key {
     return [self objectForKey:key];
 }
 
+- (NSArray*)allKeys {
```
`NSArray *`

SatbirTanda commented on SmartDeviceLink/NSMutableDictionary+Store.m:
```diff
@@ -52,6 +52,13 @@ - (nullable id)sdl_objectForName:(SDLRPCParameterName)name ofClass:(Class)classT
     } else {
+        if ((classType == NSString.class) && [obj respondsToSelector:@selector(stringValue)]) {
+            // fix an issue when JSON treats number values as NSNumber and a string expected
+            NSString * numString = [obj stringValue];
```
Stick to `<<Type>> *`

SatbirTanda commented on Example Apps/Example Video ObjC/ProxyManager.m:
```diff
@@ -0,0 +1,490 @@
+- (void)sdlex_setupConfigurationWithLifecycleConfiguration:(SDLLifecycleConfiguration *)lifecycleConfiguration {
+    if (self.sdlManager != nil) {
+        // Manager already created, just start it again.
+        //TODO: the ip might change but we still start it with the old config
```
Remove or implement all TODOs

yoooriii commented:
found a few inconsistencies in incoming SDLVideoStreamingCapability:
the following fields come empty:
scale, supportedFormats, diagonalScreenSize, pixelPerInch, preferredResolution, maxBitrate, hapticSpatialDataSupported.
Try to take absent data from the previous video capability.

yoooriii commented:
filter min/max resolution logic requires an update

yoooriii commented:
The develop has changed a lot, need to resolve conflicts and make the app work again, same goes with tests.

yoooriii commented:
Note: there is a corner case in the filter settings **supportedLandscapeStreamingRange** and **supportedPortraitStreamingRange**. The thing is, a square resolution may belong to either one since it is not landscape nor portrait. To handle such a case I can add a 3rd filter setting which is: **supportedSquareStreamingRange**.

yoooriii commented:
Update: square resolution belongs to both landscape and portrait ones and should pass either one (not necessarily both, OR logic). In other words it gets filtered out if it passes neither one.

yoooriii commented:
Hello @SatbirTanda @joeljfischer, I have updated the logic and fixed unreliable tests as well (SDLStreamingVideoLifecycleManagerSpec). Could you please review?

standa1-ford commented:
Great job, please resolve all merge conflicts

standa1-ford commented:
@theresalech ready for review

yoooriii commented:
Hello @standa1-ford. I have resolved the conflicts could you please proceed?

standa1-ford commented:
hi @yoooriii looks like one file `bson_c_lib` is still seeing a conflict

standa1-ford commented:
hi @yoooriii its approved

yoooriii commented:
Hello @joeljfischer , could you please review and merge? Thank you.

jordynmackool commented:
Hi @yoooriii, this feature is planned for the upcoming April 14, 2021 Release. We will review as time allows and reach out if we have questions. Thank you!

joeljfischer commented:
Hi @yoooriii, this PR has merge conflicts that need to be resolved, is marked *not ready* for review, claims to make major API changes, which should not be happening per the proposal, changes 87 files including many that have no business being changed, and the testing information was not filled out or was removed. All of this must be fixed before it can be considered for review.

yoooriii commented:
What was done in the latest 17 commits:
Merged with develop branch and conflicts were resolved.
The example iOS application was removed as not related to the PR itself.
The code base was cleaned up to minimise the impact in the project which decreased the number of files affected from 87 to 60.
Test cases were added and updated to increase the code coverage up to 80% ++.
The main logic remains intact.
Note: the following unit test **SDLVoiceCommandManagerSpec** was disabled as the only failing test not directly related to the PR.

yoooriii commented:
Hello @mrapitis, could you please review the PR?

a-prykho commented:
Hi @SatbirTanda ,could you please review the latest changes and approve?

standa1-ford commented on SmartDeviceLink/private/SDLCarWindow.m:
```diff
@@ -103,6 +108,14 @@ - (void)syncFrame {
     }
 }
 
+- (void)updateVdeoStreamingCapability:(SDLVideoStreamingCapability *)videoStreamingCapability {
```
```suggestion
- (void)updateVideoStreamingCapability:(SDLVideoStreamingCapability *)videoStreamingCapability {
```

standa1-ford commented on SmartDeviceLink/private/SDLCarWindow.m:
```diff
@@ -72,7 +72,12 @@ - (void)syncFrame {
 
-    CGRect bounds = self.rootViewController.view.bounds;
+    const CGRect bounds = self.streamManager.videoScaleManager.appViewportFrame;
+    if (1 > bounds.size.width) {
```
```suggestion
    if (bounds.size.width < 1) {
```

standa1-ford commented on SmartDeviceLink/private/SDLCarWindow.m:
```diff
@@ -191,75 +205,18 @@ + (nullable CVPixelBufferRef)sdl_pixelBufferForImageRef:(CGImageRef)imageRef usi
 - (void)sdl_applyDisplayDimensionsToRootViewController:(UIViewController *)rootViewController {
-    if (self.streamManager.videoScaleManager.appViewportFrame.size.width == 0) {
+    const CGSize displSize = self.streamManager.videoScaleManager.displayViewportResolution;
+    if (1 > displSize.width) {
```
```suggestion
    if (displSize.width < 1) {
```

standa1-ford commented on SmartDeviceLink/private/SDLStreamingVideoLifecycleManager.m:
```diff
@@ -644,6 +685,212 @@ - (void)sdl_hmiStatusDidChange:(SDLRPCNotificationNotification *)notification {
+    SDLVideoStreamingCapability *videoCapability = (nil == videoCapabilityUpdated) ? [self sdl_defaultVideoCapability] : videoCapabilityUpdated;
+    NSArray<SDLVideoStreamingCapability *>* capabilityMatches = [self matchVideoCapability:videoCapability];
+
+    if (0 == capabilityMatches.count) {
```
```suggestion
    if (capabilityMatches.count == 0) {
```

standa1-ford commented on SmartDeviceLink/private/SDLStreamingVideoLifecycleManager.m:
```diff
@@ -644,6 +685,212 @@ - (void)sdl_hmiStatusDidChange:(SDLRPCNotificationNotification *)notification {
+    }
+
+    SDLVideoStreamingCapability *matchedVideoCapability = capabilityMatches.firstObject;
+    if (1 < capabilityMatches.count) {
```
```suggestion
    if (capabilityMatches.count > 1) {
```

standa1-ford commented on SmartDeviceLink/private/SDLStreamingVideoLifecycleManager.m:
```diff
@@ -370,61 +413,23 @@ - (void)didEnterStateVideoStreamStopped {
-
-        [self sdl_sendVideoStartService];
-    }];
+    if (self.shouldAutoResume && (nil != self.videoStreamingCapabilityUpdated)) {
```
```suggestion
    if (self.shouldAutoResume && (self.videoStreamingCapabilityUpdated != nil)) {
```

standa1-ford commented on SmartDeviceLink/private/SDLStreamingVideoLifecycleManager.m:
```diff
@@ -644,6 +685,212 @@ - (void)sdl_hmiStatusDidChange:(SDLRPCNotificationNotification *)notification {
     }
 }
 
+- (void)subscribeForNotifications:(NSNotificationCenter*)notificationCenter {
```
```suggestion
- (void)subscribeForNotifications:(NSNotificationCenter *)notificationCenter {
```

standa1-ford commented on SmartDeviceLink/private/SDLStreamingVideoLifecycleManager.m:
```diff
@@ -644,6 +685,212 @@ - (void)sdl_hmiStatusDidChange:(SDLRPCNotificationNotification *)notification {
+
+- (void)sdl_applyVideoCapabilityWhileStarting:(nullable SDLVideoStreamingCapability *)videoCapabilityUpdated {
+    SDLVideoStreamingCapability *videoCapability = (nil == videoCapabilityUpdated) ? [self sdl_defaultVideoCapability] : videoCapabilityUpdated;
+    NSArray<SDLVideoStreamingCapability *>* capabilityMatches = [self matchVideoCapability:videoCapability];
```
```suggestion
    NSArray<SDLVideoStreamingCapability *> *capabilityMatches = [self matchVideoCapability:videoCapability];
```

standa1-ford commented on SmartDeviceLink/private/SDLStreamingVideoLifecycleManager.m:
```diff
@@ -644,6 +685,212 @@ - (void)sdl_hmiStatusDidChange:(SDLRPCNotificationNotification *)notification {
+    if (!videoCapability.hapticSpatialDataSupported) {
+        videoCapability.hapticSpatialDataSupported = self.videoStreamingCapability.hapticSpatialDataSupported;
+    }
+    if (0 == videoCapability.supportedFormats.count) {
```
```suggestion
    if (videoCapability.supportedFormats.count == 0) {
```

standa1-ford commented on SmartDeviceLink/private/SDLStreamingVideoLifecycleManager.m:
```diff
@@ -644,6 +685,212 @@ - (void)sdl_hmiStatusDidChange:(SDLRPCNotificationNotification *)notification {
+        return @[];
+    }
+
+    NSArray <SDLVideoStreamingCapability*>* allCapabilities = [videoStreamingCapability allVideoStreamingCapabilitiesPlain];
```
```suggestion
    NSArray <SDLVideoStreamingCapability*> *allCapabilities = [videoStreamingCapability allVideoStreamingCapabilitiesPlain];
```

standa1-ford commented on SmartDeviceLink/private/SDLStreamingVideoLifecycleManager.m:
```diff
@@ -644,6 +685,212 @@ - (void)sdl_hmiStatusDidChange:(SDLRPCNotificationNotification *)notification {
+
+    NSArray <SDLVideoStreamingCapability*>* allCapabilities = [videoStreamingCapability allVideoStreamingCapabilitiesPlain];
+    NSMutableArray *matchCapabilities = [NSMutableArray arrayWithCapacity:allCapabilities.count];
+    for (SDLVideoStreamingCapability* nextCapability in allCapabilities) {
```
```suggestion
    for (SDLVideoStreamingCapability *nextCapability in allCapabilities) {
```

standa1-ford commented on SmartDeviceLink/private/SDLStreamingVideoLifecycleManager.m:
```diff
@@ -797,24 +1025,59 @@ - (void)sdl_sendVideoStartService {
 
-#pragma mark Setters / Getters
+- (void)sdl_applyVideoCapability:(SDLVideoStreamingCapability *)capability {
+    assert(nil != capability);
```
```suggestion
    assert(capability != nil);
```
I'm not sure if asserting is a graceful way of handling the nil case here

yoooriii replied:
It does not handle nil but it fails early and besides nil checks are implemented in caller methods.

standa1-ford commented on SmartDeviceLink/public/SDLFunctionID.m:
```diff
@@ -117,7 +117,8 @@ - (instancetype)init {
                          @32786: SDLRPCFunctionNameOnAppServiceData,
                          @32787: SDLRPCFunctionNameOnSystemCapabilityUpdated,
-                         @32788: SDLRPCFunctionNameOnSubtleAlertPressed,
+//                         @32788: SDLRPCFunctionNameOnSubtleAlertPressed,  //Warning: duplicate f-id
```
if this is not used, we can remove it

yoooriii replied:
<element name="OnAppCapabilityUpdatedID" value="32791" hexvalue="8017" since="7.1"/>
The function id had a conflict with another one, looks like it is resolved in MOBILE_API.xml

standa1-ford commented on SmartDeviceLink/public/SDLImageResolution.m:
```diff
@@ -39,6 +46,53 @@ - (void)setResolutionHeight:(NSNumber<SDLInt> *)resolutionHeight {
+- (float)normalizedAspectRatio {
+    const float width = self.resolutionWidth.floatValue;
+    const float height = self.resolutionHeight.floatValue;
+    return (0 == width || 0 == height) ? 0 : fmaxf(width, height)/fminf(width, height);
```
```suggestion
    return (width == 0 || height == 0) ? 0 : fmaxf(width, height)/fminf(width, height);
```

standa1-ford commented on SmartDeviceLink/public/SDLImageResolution.m:
```diff
@@ -39,6 +46,53 @@ - (void)setResolutionHeight:(NSNumber<SDLInt> *)resolutionHeight {
+        return SDLImageResolutionKindUndefined;
+    }
+    const float ratio = self.resolutionWidth.floatValue/self.resolutionHeight.floatValue;
+    const float ratio2 = ratio * ratio;
```
```suggestion
    const float ratioSquared = ratio * ratio;
```

standa1-ford commented on SmartDeviceLink/public/SDLImageResolution.m:
```diff
@@ -39,6 +46,53 @@ - (void)setResolutionHeight:(NSNumber<SDLInt> *)resolutionHeight {
+    const float ratio = self.resolutionWidth.floatValue/self.resolutionHeight.floatValue;
+    const float ratio2 = ratio * ratio;
+    const float tolerance = 0.001f;
+    if (ratio2 < 1.0 - tolerance) {
```
```suggestion
    if (ratioSquared < 1.0 - tolerance) {
```

standa1-ford commented on SmartDeviceLink/public/SDLImageResolution.m:
```diff
@@ -39,6 +46,53 @@ - (void)setResolutionHeight:(NSNumber<SDLInt> *)resolutionHeight {
+    if (ratio2 < 1.0 - tolerance) {
+        return SDLImageResolutionKindPortrait;
+    }
+    if (ratio2 > 1.0 + tolerance) {
```
```suggestion
    if (ratioSquared > 1.0 + tolerance) {
```

standa1-ford commented on SmartDeviceLink/public/SDLImageResolution.m:
```diff
@@ -39,6 +46,53 @@ - (void)setResolutionHeight:(NSNumber<SDLInt> *)resolutionHeight {
+    return SDLImageResolutionKindSquare;
+}
+
+#define EQUAL_NUM(property) ((nil == self.property && nil == res2.property) ? YES : [self.property isEqualToNumber:res2.property])
```
```suggestion
#define EQUAL_NUM(property) ((self.property == nil && res2.property == nil) ? YES : [self.property isEqualToNumber:res2.property])
```

It looks like this its defined and used multiple times

yoooriii replied:
It might be defined a few times though every time it is undefined after being used

standa1-ford commented on SmartDeviceLink/public/SDLImageResolution.m:
```diff
@@ -39,6 +46,53 @@ - (void)setResolutionHeight:(NSNumber<SDLInt> *)resolutionHeight {
+    if (![object isKindOfClass:self.class]) {
+        return NO;
+    }
+    typeof(self) res2 = object;
```
is this used?

standa1-ford commented on SmartDeviceLink/public/SDLMsgVersion.m:
```diff
@@ -12,6 +12,17 @@
 
++ (nullable instancetype)versionWithString:(NSString *)versionString {
+    NSArray<NSString *> *components = [versionString componentsSeparatedByString:@"."];
+    if (3 != components.count) {
```
```suggestion
    if (components.count != 3) {
```

standa1-ford commented on SmartDeviceLink/public/SDLSupportedStreamingRange.m:
```diff
@@ -0,0 +1,88 @@
+    }
+
+    BOOL isInRange = YES;
+    if (1.f <= self.minimumAspectRatio) {
```
```suggestion
    if (self.minimumAspectRatio >= 1.f) {
```

standa1-ford commented on SmartDeviceLink/public/SDLSupportedStreamingRange.m:
```diff
@@ -0,0 +1,88 @@
+    if (1.f <= self.minimumAspectRatio) {
+        isInRange = (aspectRatio >= self.minimumAspectRatio);
+    }
+    if (isInRange && (1.f <= self.maximumAspectRatio)) {
```
```suggestion
    if (isInRange && (self.maximumAspectRatio >= 1.f)) {
```

standa1-ford commented on SmartDeviceLink/public/SDLSupportedStreamingRange.m:
```diff
@@ -0,0 +1,88 @@
+}
+
+- (NSString *)description {
+    return [NSString stringWithFormat:@"%@: ratio-min/max:[%2.2f/%2.2f], min-diagonal:%2.2f, resolution-min/max:[%@/%@]", NSStringFromClass(self.class), self.minimumAspectRatio, self.maximumAspectRatio, self.minimumDiagonal, self.minimumResolution.stringValue, self.maximumResolution.stringValue];
```
is there a way you can break each component of the formatted string into standalone strings and then concat them?

yoooriii replied:
yes sure

standa1-ford commented on SmartDeviceLink/public/SDLSystemCapabilityManager.m:
```diff
@@ -87,6 +87,7 @@ @interface SDLSystemCapabilityManager ()
 
 @end
 
+
```
```suggestion
```

standa1-ford commented on SmartDeviceLink/public/SDLSystemCapabilityManager.m:
```diff
@@ -560,8 +558,7 @@ - (BOOL)subscribeToCapabilityType:(SDLSystemCapabilityType)type withObserver:(id
 
-    id<NSObject> subscribedObserver = [self sdl_subscribeToCapabilityType:type observerObject:observerObject];
-    return subscribedObserver == nil ? NO : YES;
+    return nil != [self sdl_subscribeToCapabilityType:type observerObject:observerObject];
```
```suggestion
    return [self sdl_subscribeToCapabilityType:type observerObject:observerObject] != nil;
```

standa1-ford commented on SmartDeviceLink/public/SDLSystemCapabilityManager.m:
```diff
@@ -575,12 +572,12 @@ - (BOOL)subscribeToCapabilityType:(SDLSystemCapabilityType)type withObserver:(id
-        [SDLGlobals runSyncOnSerialSubQueue:self.readWriteQueue block:^{
-            self.capabilityObservers[type] = [NSMutableArray arrayWithObject:observerObject];
-        }];
+    if (0 == observersCount) {
```
```suggestion
    if (observersCount == 0) {
```

standa1-ford commented on SmartDeviceLink/public/SDLSystemCapabilityManager.m:
```diff
@@ -735,7 +755,8 @@ - (void)sdl_registerResponse:(SDLRPCResponseNotification *)notification {
     self.prerecordedSpeechCapabilities = response.prerecordedSpeech;
-    self.vrCapability = (response.vrCapabilities.count > 0 && [response.vrCapabilities.firstObject isEqualToEnum:SDLVRCapabilitiesText]) ? YES : NO;
+    NSArray<SDLVRCapabilities> *vrCapabilities = response.vrCapabilities;
+    self.vrCapability = (0 < vrCapabilities.count && [vrCapabilities.firstObject isEqualToEnum:SDLVRCapabilitiesText]);
```
```suggestion
    self.vrCapability = (vrCapabilities.count > 0 && [vrCapabilities.firstObject isEqualToEnum:SDLVRCapabilitiesText]);
```

standa1-ford commented on SmartDeviceLink/public/SDLVideoStreamingCapability.h:
```diff
@@ -84,6 +84,19 @@ NS_ASSUME_NONNULL_BEGIN
+
+// this returns a copy array of all capabilities including itself but with no recursion
+// in the result objects the .additionalVideoStreamingCapabilities will be nil
+- (NSArray <SDLVideoStreamingCapability*>*)allVideoStreamingCapabilitiesPlain;
```
```suggestion
- (NSArray <SDLVideoStreamingCapability*> *)allVideoStreamingCapabilitiesPlain;
```

standa1-ford commented on SmartDeviceLink/public/SDLVideoStreamingCapability.m:
```diff
@@ -90,6 +91,113 @@ - (void)setScale:(nullable NSNumber<SDLFloat> *)scale {
+    return aCopy;
+}
+
+- (NSArray <SDLVideoStreamingCapability*>*)allVideoStreamingCapabilitiesPlain {
```
```suggestion
- (NSArray <SDLVideoStreamingCapability*> *)allVideoStreamingCapabilitiesPlain {
```

standa1-ford commented on SmartDeviceLink/public/SDLVideoStreamingCapability.m:
```diff
@@ -90,6 +91,113 @@ - (void)setScale:(nullable NSNumber<SDLFloat> *)scale {
+}
+
+- (SDLImageResolution *)makeImageResolution {
+    const float scale = (nil == self.scale) ? 1.0 : self.scale.floatValue;
```
```suggestion
    const float scale = (self.scale == nil) ? 1.0 : self.scale.floatValue;
```

standa1-ford commented on SmartDeviceLink/public/SDLVideoStreamingCapability.m:
```diff
@@ -90,6 +91,113 @@ - (void)setScale:(nullable NSNumber<SDLFloat> *)scale {
+    return resolutions;
+}
+
+#define EQUAL_RES ((nil == self.preferredResolution && nil == cap2.preferredResolution) ? YES : [self.preferredResolution isEqual:cap2.preferredResolution])
```
```suggestion
#define EQUAL_RES ((self.preferredResolution == nil && cap2.preferredResolution == nil) ? YES : [self.preferredResolution isEqual:cap2.preferredResolution])
```

standa1-ford commented on SmartDeviceLink/public/SDLVideoStreamingCapability.m:
```diff
@@ -90,6 +91,113 @@ - (void)setScale:(nullable NSNumber<SDLFloat> *)scale {
+}
+
+#define EQUAL_RES ((nil == self.preferredResolution && nil == cap2.preferredResolution) ? YES : [self.preferredResolution isEqual:cap2.preferredResolution])
+#define EQUAL_NUM(property) ((nil == self.property && nil == cap2.property) ? YES : [self.property isEqualToNumber:cap2.property])
```
```suggestion
#define EQUAL_NUM(property) ((self.property == nil && cap2.property == nil) ? YES : [self.property isEqualToNumber:cap2.property])
```

standa1-ford commented on SmartDeviceLink/public/SDLVideoStreamingCapability.m:
```diff
@@ -90,6 +91,113 @@ - (void)setScale:(nullable NSNumber<SDLFloat> *)scale {
+
+- (NSString *)description {
+    NSMutableString *formats = [NSMutableString string];
+    for (SDLVideoStreamingFormat * f in self.supportedFormats) {
```
```suggestion
    for (SDLVideoStreamingFormat *f in self.supportedFormats) {
```

standa1-ford commented on SmartDeviceLink/public/SDLVideoStreamingCapability.m:
```diff
@@ -90,6 +91,113 @@ - (void)setScale:(nullable NSNumber<SDLFloat> *)scale {
+    for (SDLVideoStreamingFormat * f in self.supportedFormats) {
+        [formats appendFormat:@"%@; ", f];
+    }
+    return [NSString stringWithFormat:@"<%@:%p>{\n\tsupportedFormats:%@\n\tpreferredResolution:%@\n\tmaxBitrate:%@\n\thapticSpatialDataSupported:%@\n\tdiagonalScreenSize:%@\n\tpixelPerInch:%@\n\tscale:%@\n\tadditionalVideoStreamingCapabilities[%d]:%@ }", NSStringFromClass(self.class), self, formats, self.preferredResolution, self.maxBitrate, self.hapticSpatialDataSupported.boolValue ? @"YES" : @" ...
```
is there a way you can break each component of the formatted string into stand-alone string variables and then concat them?

yoooriii replied:
yes sure there is

standa1-ford commented on SmartDeviceLinkTests/DevAPISpecs/SDLStreamingVideoLifecycleManagerSpec.m:
```diff
@@ -33,103 +35,424 @@
 
+// expose private methods to the test suite
+@interface SDLVideoStreamingCapability (test)
+- (NSArray <SDLVideoStreamingCapability*>*)allVideoStreamingCapabilitiesPlain;
```
```suggestion
- (NSArray <SDLVideoStreamingCapability*> *)allVideoStreamingCapabilitiesPlain;
```

a-prykho commented:
Dear @jordynmackool  @joeljfischer this PR is ready for Livio review.

jordynmackool commented:
@standa1-ford is looks like changes were made to the PR after your approval - please confirm this is ready for Livio review.

standa1-ford commented:
@jordynmackool approved

NicoleYarroch commented:
@yoooriii PR #1878 was approved and merged with the develop branch. Can you fix the merge conflicts with the develop branch before I start the review of this PR?

yoooriii commented:
Hello @NicoleYarroch. I have updated the PR and resolved conflicts. Could you please proceed? Thank you.

NicoleYarroch reviewed (changes requested):
I requested some general cleanup and made some requests to change class and method names to match the RPC spec.

NicoleYarroch commented:
@yoooriii Please tag me when you are ready for Livio to do another review. Thanks!

yoooriii commented on SmartDeviceLink/private/SDLLifecycleManager.h:
```diff
@@ -86,6 +86,11 @@ typedef void (^SDLManagerReadyBlock)(BOOL success, NSError *_Nullable error);
 
 @property (strong, nonatomic) NSOperationQueue *rpcOperationQueue;
 
+/**
+ * The version number of the SDL V4 interface, string representation
+ * Optional
+ */
+@property (copy, nonatomic, nullable) NSString *sdlMsgVersionString;
```
The thing is, the head unit behaves (responses) differently depending on this version. The client (mobile) application can optionally set this value. I do not know if it actually needed or not.

jordynmackool commented:
@yoooriii, Livio is standing by ready to re-review, please let us know when this PR is updated based on the review comments. Thanks!

yoooriii commented:
Hello @NicoleYarroch. I have updated the PR and depending tests as well. Could you please proceed? Thank you.

NicoleYarroch commented on SmartDeviceLink/private/SDLStreamingVideoLifecycleManager.m:
```diff
@@ -659,6 +687,216 @@ - (void)sdl_hmiStatusDidChange:(SDLRPCNotificationNotification *)notification {
+    }
+
+    // create and send SDLOnAppCapabilityUpdated notification
+    SDLAppCapability *appCapability = [[SDLAppCapability alloc] initWithAppCapabilityType:SDLAppCapabilityTypeVideoStreaming videoStreamingCapability:matchedVideoCapability];
```
The proposal was not very clear on what properties should get set on `OnAppCapabilityUpdated.appCapability. videoStreamingCapability` so we have decided the mobile libraries should not set the root `videoStreamingCapability` but only the `additionalVideoStreamingCapabilities` as outlined below:

1. Mobile checks both the root `videoStreamingCapability` and the `additionalVideoStreamingCapabilities` to see which ones it can support. 
2. Mobile sends an `OnAppCapabilityUpdated` with only the `VideoStreamingCapability.additionalVideoStreamingCapabilities` parameter set. The  `additionalVideoStreamingCapabilities` should be set with all the capabilities it can support as determined by step 1.

NicoleYarroch commented on SmartDeviceLink/private/SDLCarWindow.m:
```diff
@@ -162,7 +175,8 @@ - (void)setRootViewController:(nullable UIViewController *)rootViewController {
 
 #pragma mark - Private Helpers
-+ (nullable CVPixelBufferRef)sdl_pixelBufferForImageRef:(CGImageRef)imageRef usingPool:(CVPixelBufferPoolRef)pool {
+// memory management 'create': release the result object when done (create* - follow the naming convention)
```
```suggestion
```

NicoleYarroch commented on SmartDeviceLinkTests/SDLStreamingVideoScaleManagerSpec.m:
```diff
@@ -40,22 +41,24 @@
-            videoScaleManager.scale = 0.3;
-            CGRect expectedRect = CGRectMake(0, 0, testScreenSize.width, testScreenSize.height);
-            CGRect testRect = videoScaleManager.appViewportFrame;
+        it(@"should scale the frame correctly with the scale 0.3", ^{
+            const float scale = 0.3f;
+            videoScaleManager.scale = scale;
+            const CGRect expectedRect = CGRectMake(0, 0, roundf(testScreenSize.width/scale), roundf(testScreenSize.height/scale));
+            const CGRect testRect = videoScaleManager.appViewportFrame;
```
I don't think this test is needed as the RPC spec says that scale has a min value of 1.0

https://github.com/smartdevicelink/rpc_spec/blob/8b8c996c738711c4235c8574c179354cbf8982e2/MOBILE_API.xml#L3293

NicoleYarroch reviewed (changes requested):
@yoooriii To help speed up the review process can you please make sure that the branch builds and edge cases have been tested with the sdl_core branch before asking for a review. Thanks!

Please tag me when you are ready for Livio to do a review again.

yoooriii commented on SmartDeviceLink/private/SDLStreamingVideoLifecycleManager.m:
```diff
@@ -659,6 +687,216 @@ - (void)sdl_hmiStatusDidChange:(SDLRPCNotificationNotification *)notification {
+#pragma mark - Subscribed notifications
+
+- (void)sdl_videoStreamingCapabilityDidUpdate:(SDLSystemCapability *)systemCapability {
+    SDLVideoStreamingCapability *videoCapability = systemCapability.videoStreamingCapability;
```
Revert the old method:
`[self sdl_requestVideoCapabilities:^(SDLVideoStreamingCapability * _Nullable capability) {...}]`
It fixes the issue.

yoooriii commented:
Hello @NicoleYarroch. Thank you for review. I have updated the PR and made sure it works. Could you please review it? Thank you.

NicoleYarroch commented on SmartDeviceLink/private/SDLStreamingVideoLifecycleManager.m:
```diff
@@ -329,6 +354,21 @@ - (void)didEnterStateAppActive {
     }
 }
 
+- (void)suspendVideo {
```
```suggestion
- (void)sdl_suspendVideo {
```

NicoleYarroch commented on SmartDeviceLink/private/SDLStreamingVideoLifecycleManager.m:
```diff
@@ -304,21 +329,21 @@ - (void)didEnterStateAppInactive {
+    [self resumeVideo];
+}
+
+- (void)resumeVideo {
```
```suggestion
- (void)sdl_resumeVideo {
```

NicoleYarroch commented on SmartDeviceLink/private/SDLStreamingVideoLifecycleManager.m:
```diff
@@ -856,6 +1167,18 @@ - (BOOL)isStreamingSupported {
     return (self.systemCapabilityManager != nil) ? [self.systemCapabilityManager isCapabilitySupported:SDLSystemCapabilityTypeVideoStreaming] : YES;
 }
 
+- (void)setPreferredResolutions:(NSArray<SDLImageResolution *> *)resolutions {
+    _preferredResolutions = resolutions;
+    _preferredResolutionIndex = 0;
+    [self didChangeValueForKey:@"preferredResolutions"];
+}
+
+- (void)setPreferredFormats:(NSArray<SDLVideoStreamingFormat *> *)formats {
+    _preferredFormats = formats;
+    _preferredFormatIndex = 0;
+    [self didChangeValueForKey:@"preferredFormats"];
+}
```
Are these setters needed? There doesn't appear to be any place in the code where `preferredResolutions` or `preferredFormats` is observed.

NicoleYarroch commented on SmartDeviceLink/public/SDLImageResolution.h:
```diff
@@ -1,15 +1,15 @@
 //
 
-
+#import <CoreGraphics/CGGeometry.h>
```
```suggestion
```

NicoleYarroch commented on SmartDeviceLink/public/SDLStreamingMediaConfiguration.h:
```diff
@@ -82,6 +83,10 @@ typedef NS_ENUM(NSUInteger, SDLCarWindowRenderingType) {
  */
 @property (assign, nonatomic) BOOL allowMultipleViewControllerOrientations;
 
+@property (strong, nonatomic, nullable) SDLVideoStreamingRange *supportedLandscapeStreamingRange;
+@property (strong, nonatomic, nullable) SDLVideoStreamingRange *supportedPortraitStreamingRange;
+@property (weak, nonatomic, nullable) id<SDLStreamingVideoDelegate> delegate;
```
Please add documentation

NicoleYarroch reviewed (changes requested):
Left some suggestions.

NicoleYarroch commented:
@yoooriii Please tag me when you are ready for Livio to do another review. Thanks!

yoooriii commented on SmartDeviceLink/private/SDLStreamingVideoLifecycleManager.m:
```diff
@@ -659,6 +690,235 @@ - (void)sdl_hmiStatusDidChange:(SDLRPCNotificationNotification *)notification {
+    if (videoCapability) {
+        [self sdl_useVideoCapability:videoCapability];
+    } else {
+        typeof(self) weakSelf = self;
+        [self sdl_requestVideoCapabilities:^(SDLVideoStreamingCapability * _Nullable capability) {
+            SDLLogD(@"Received video capability response");
+            SDLLogV(@"Capability: %@", capability);
+            [weakSelf sdl_useVideoCapability:capability];
+        }];
```
Update: in the last line videoCapability is nil. 
```
#pragma clang diagnostic push
#pragma clang diagnostic ignored "-Wnonnull"
        [self sdl_useVideoCapability:nil];
#pragma clang diagnostic pop

```

yoooriii commented on SmartDeviceLink/private/SDLVideoStreamingCapability + StreamingVideoExtensions.h:
```diff
@@ -0,0 +1,29 @@
+
+@interface SDLVideoStreamingCapability (StreamingVideoExtensions)
+
+// this returns a copy array of all capabilities including itself but with no recursion
```
Yes it does. There are 2 kinds of recursion, 1 - the recursion in the struct (it contains references to other instances of the same struct) and 2- in the method. The method finds all recursive struct instances and places them in a plain collection, the result items in the collection have no recursion.

yoooriii replied:
To make things clear I removed the "but with no recursion" part.

yoooriii commented on SmartDeviceLink/private/SDLStreamingVideoLifecycleManager.m:
```diff
@@ -659,6 +690,235 @@ - (void)sdl_hmiStatusDidChange:(SDLRPCNotificationNotification *)notification {
+
+- (void)sdl_applyVideoCapabilityWhileStarting:(nullable SDLVideoStreamingCapability *)videoCapabilityUpdated {
+    SDLVideoStreamingCapability *videoCapability = (nil == videoCapabilityUpdated) ? [self sdl_defaultVideoCapability] : videoCapabilityUpdated;
+    NSArray<SDLVideoStreamingCapability *> *capabilityMatches = [self matchVideoCapability:videoCapability];
```
Fixed. If the array is empty or nil then a new array is created containing either the root item (a shallow copy) or a default one if none is available.

yoooriii commented:
Hello @NicoleYarroch. I have fixed the PR as requested in comments. Could you please review? Thank you.

NicoleYarroch commented on SmartDeviceLink/private/SDLStreamingVideoLifecycleManager.m:
```diff
@@ -659,6 +690,242 @@ - (void)sdl_hmiStatusDidChange:(SDLRPCNotificationNotification *)notification {
+    }
+    [self sdl_applyVideoCapability:matchedVideoCapability];
+
+    if (self.delegate) {
+        __weak typeof(self) weakSelf = self;
+        const CGSize displaySize = matchedVideoCapability.makeImageResolution.makeSize;
+        dispatch_async(dispatch_get_main_queue(), ^{
+            [weakSelf.delegate videoStreamingSizeDidUpdate:displaySize];
+        });
+    }
```
I wonder if notifying the observer should happen after we get the video start service ACK? Otherwise we are telling them to change the display size before we even know if the module will allow video streaming.

yoooriii replied:
The delegate must know if the size updated. It is up to the client app to take an action or not based on the call.

NicoleYarroch commented on SmartDeviceLink/private/SDLStreamingVideoLifecycleManager.m:
```diff
@@ -659,6 +690,242 @@ - (void)sdl_hmiStatusDidChange:(SDLRPCNotificationNotification *)notification {
+        matchedVideoCapability.additionalVideoStreamingCapabilities = @[];
+    }
+
+    // take formats from 'parent' since onse may be absent in children
```
```suggestion
    // take formats from 'parent' if absent in children
```

NicoleYarroch commented on SmartDeviceLink/private/SDLVideoStreamingCapability + StreamingVideoExtensions.h:
```diff
@@ -14,16 +14,33 @@ NS_ASSUME_NONNULL_BEGIN
 - (SDLImageResolution *)makeImageResolution;
 
+/**
+ It traverces through all capabilities and collects all image resolutions into an array respecting the scale factor
```
```suggestion
 It traverses through all capabilities and collects all image resolutions into an array respecting the scale factor
```

NicoleYarroch commented on SmartDeviceLink/private/SDLVideoStreamingCapability + StreamingVideoExtensions.h:
```diff
@@ -14,16 +14,33 @@ NS_ASSUME_NONNULL_BEGIN
 - (NSArray<SDLImageResolution *> *)allImageResolutionsScaled;
 
+/**
+ It traverces through all capabilities and collects all image resolutions into an array ignoring the scale factor
```
```suggestion
 It traverses through all capabilities and collects all image resolutions into an array ignoring the scale factor
```

NicoleYarroch reviewed (changes requested):
@yoooriii Please tag me when you are ready for Livio to do a review. Thanks!

yoooriii commented on SmartDeviceLink/private/SDLStreamingVideoLifecycleManager.m:
```diff
@@ -659,6 +690,242 @@ - (void)sdl_hmiStatusDidChange:(SDLRPCNotificationNotification *)notification {
+    SDLVideoStreamingCapability *notifyCapability = [[SDLVideoStreamingCapability alloc] init];
+    notifyCapability.additionalVideoStreamingCapabilities = supportedVideoStreamingCapabilities ?: @[];
+    SDLAppCapability *appCapability = [[SDLAppCapability alloc] initWithAppCapabilityType:SDLAppCapabilityTypeVideoStreaming videoStreamingCapability:notifyCapability];
+    SDLOnAppCapabilityUpdated *notification = [[SDLOnAppCapabilityUpdated alloc] initWithAppCapability:appCapability];
```
Added a bool flag shouldSendOnAppCapabilityUpdated to control this flow

yoooriii commented:
Hello @NicoleYarroch. I have updated the PR. Could you please review it? Thank you.

NicoleYarroch reviewed (changes requested):
I made some suggestions.

yoooriii commented:
Hello @NicoleYarroch. I have updated the PR, could you review please?

I updated the delegate call as well.
https://github.com/smartdevicelink/sdl_ios/pull/1756#discussion_r579333855

 Thank you.

NicoleYarroch commented on SmartDeviceLink/private/SDLStreamingVideoLifecycleManager.m:
```diff
@@ -454,6 +453,15 @@ - (void)didEnterStateVideoStreamReady {
 
     [[NSNotificationCenter defaultCenter] postNotificationName:SDLVideoStreamDidStartNotification object:nil];
 
+    if (self.shouldUpdateDelegateOnSizeChange && self.delegate != nil) {
```
Just curious why there is a `shouldUpdateDelegateOnSizeChange` flag? I believe `didEnterStateVideoStreamReady` should only be called once the module has ACK the request the start the video stream so I don't think there is any danger of `didEnterStateVideoStreamReady` getting called more than once. Is there some edge case I am not thinking of in relation to this?

yoooriii replied:
The `SDLVideoStreamingCapability` may change in 2 cases (2 states): `SDLVideoStreamManagerStateStarting` and `SDLVideoStreamManagerStateReady`. The notification will be sent only in the 1st case and skipped in the 2nd. 
If it is wrong I can remove this check.

NicoleYarroch replied:
I see what is going on now. If the video is in the suspended state but has then been resumed we don't want to send a duplicate `videoStreamingSizeDidUpdate` notification to the observer.

NicoleYarroch commented on SmartDeviceLinkTests/SDLStreamingVideoScaleManagerSpec.m:
```diff
@@ -118,7 +113,7 @@
 
          it(@"should scale the rectangle correctly with a scale < 1", ^{
```
This test is failing

NicoleYarroch commented on SmartDeviceLinkTests/ProxySpecs/SDLHapticManagerSpec.m:
```diff
@@ -550,15 +569,17 @@ @interface SDLFocusableItemLocator ()
 
-                     compareScaledRectangle(sdlRect, viewRect1, 1.0);
+                     if (!compareScaledRectangle(sdlRect, viewRect1, testUpdatedScale)) {
+                         failWithMessage(([NSString stringWithFormat:@"rects are not equal {%@ vs %@}", sdlRect, NSStringFromCGRect(viewRect1)]));
```
This test is failing

NicoleYarroch reviewed (changes requested):
I left a few comments. 

I noticed that this PR also makes changes to the `Submodule rpc_spec`. Can you make sure to remove those changes before your next push (Since it is a submodule, we don't want to make any changes to it)

NicoleYarroch commented:
@yoooriii Can you please tag me when you are ready for Livio to review again. Thanks!

yoooriii commented:
Hello @NicoleYarroch. I have updated the PR. Could you please review? Also could you please advice if I should remove this flag: `shouldUpdateDelegateOnSizeChange` and remove the logic connected with it.

NicoleYarroch commented on SmartDeviceLink/private/SDLStreamingVideoLifecycleManager.m:
```diff
@@ -454,6 +453,15 @@ - (void)didEnterStateVideoStreamReady {
 
     [[NSNotificationCenter defaultCenter] postNotificationName:SDLVideoStreamDidStartNotification object:nil];
 
+    if (self.shouldUpdateDelegateOnSizeChange && self.delegate != nil) {
```
```suggestion
   // Don't update the delegate if video is resuming after being suspended.  
    if (self.shouldUpdateDelegateOnSizeChange && self.delegate != nil) {
```

NicoleYarroch commented:
@yoooriii You were right, we do need the `shouldUpdateDelegateOnSizeChange` flag. I found a couple of edge cases while testing that need to get fixed. Please let me know if you need more detail to reproduce. Thanks.

yoooriii commented on SmartDeviceLink/private/SDLStreamingVideoLifecycleManager.m:
```diff
@@ -659,6 +704,240 @@ - (void)sdl_hmiStatusDidChange:(SDLRPCNotificationNotification *)notification {
+    } else if ([self.videoStreamStateMachine.currentState isEqualToEnum:SDLVideoStreamManagerStateReady]) {
+        [self sdl_applyVideoCapabilityWhenStreaming:videoCapability];
+    } else {
+        SDLLogD(@"Received video capabilities in wrong state: %@; ignore it", self.videoStreamStateMachine.currentState);
```
I have updated the method though I am afraid it was not in the proposal and I am not sure we should fix it in this PR.

yoooriii commented:
Hello @NicoleYarroch. I have updated the PR. Could you please verify it? Thank you.

NicoleYarroch commented on SmartDeviceLink/public/SDLStreamingMediaConfiguration.h:
```diff
@@ -82,6 +83,21 @@ typedef NS_ENUM(NSUInteger, SDLCarWindowRenderingType) {
  */
 @property (assign, nonatomic) BOOL allowMultipleViewControllerOrientations;
 
+/**
+ A configuration object of type SDLVideoStreamingRange. If the application supports a specific video parameters in landscape orientation like screen size etc then please provide the parameters in this struct.
+ */
```
```suggestion
/**
 Set a landscape image dimension range and/or aspect ratio range that your rootViewController supports. If the module's screen size for your app changes during streaming (i.e. to a collapsed view, split screen, preview mode or picture-in-picture), your rootViewController will be resized to the new screen size. If left unset or set to `nil`, the default is to support all lanscape streaming ranges. If you wish to disable support for streaming in landscape mode, set a `disabled` video streaming range.
 If desired, you can subscribe to screen size updates via the SDLStreamingVideoDelegate.
 @warning If you disable both the supportedLandscapeStreamingRange and supportedPortraitStreamingRange, video will not stream
 */
```

NicoleYarroch commented on SmartDeviceLink/public/SDLStreamingMediaConfiguration.h:
```diff
@@ -82,6 +83,21 @@ typedef NS_ENUM(NSUInteger, SDLCarWindowRenderingType) {
+ */
+@property (strong, nonatomic, nullable) SDLVideoStreamingRange *supportedLandscapeStreamingRange;
+
+/**
+ A configuration object of type SDLVideoStreamingRange. If the application supports a specific video parameters in portrait orientation like screen size etc then please provide the parameters in this struct.
+ */
```
```suggestion
 /**
 Set a portrait image dimension range and/or aspect ratio range that your rootViewController supports. If the module's screen size for your app changes during streaming (i.e. to a collapsed view, split screen, preview mode or picture-in-picture), your rootViewController will be resized to the new screen size. If left unset or set to `nil`, the default is to support all portrait streaming ranges. If you wish to disable support for streaming in portrait mode, set a `disabled` video streaming range.
 If desired, you can subscribe to screen size updates via the SDLStreamingVideoDelegate.
 @warning If you disable both the supportedLandscapeStreamingRange and supportedPortraitStreamingRange, video will not stream
 */
```

NicoleYarroch reviewed (changes requested):
Testing went well. I just requested some cleanup and clarification on some of the documentation and I found a few private methods that need the `sdl_` prefix (you might need to update the tests as well after updating the method names).

yoooriii commented:
Hello @NicoleYarroch. I have updated the PR and merged develop into it. Would you please take a look? Thank you.

NicoleYarroch reviewed (changes requested):
I have one small change request after doing alignment with the Java-Suite PR.

NicoleYarroch commented:
@yoooriii Please tag me when you are ready for Livio to do another review. Thanks!

yoooriii commented:
Hello @NicoleYarroch, I have added this log output. Could you please proceed? Thank you.
